kotlin/workflow-runtime/src/jmh/java/com/squareup/workflow/WorkflowNodeBenchmark.kt

```diff
@@ -0,0 +1,168 @@
+import org.openjdk.jmh.annotations.Fork
+import org.openjdk.jmh.annotations.Measurement
+import org.openjdk.jmh.annotations.Mode.AverageTime
+import org.openjdk.jmh.annotations.OutputTimeUnit
+import org.openjdk.jmh.annotations.Param
+import org.openjdk.jmh.annotations.Scope
+import org.openjdk.jmh.annotations.Setup
+import org.openjdk.jmh.annotations.State
+import org.openjdk.jmh.annotations.Warmup
+import java.util.concurrent.TimeUnit.MICROSECONDS
+import java.util.concurrent.TimeUnit.SECONDS
+
+@Fork(value = 1)
+@State(Scope.Thread)
+@BenchmarkMode(AverageTime)
+@Warmup(iterations = 3, time = 5, timeUnit = SECONDS)
+@Measurement(iterations = 3, time = 5, timeUnit = SECONDS)
+@OutputTimeUnit(MICROSECONDS)
+open class WorkflowNodeBenchmark {
+
+  /**
+   * Used to parameterize the benchmarks.
+   *
+   * @param childCount The number of children _each workflow in the tree_ should have.
+   * @param depth The depth of the workflow tree.
+   */
+  enum class TreeShape(
+    val childCount: Int,
+    val depth: Int
+  ) {
+    DEEP(childCount = 1, depth = 100),
+    BUSHY(childCount = 100, depth = 1),
+    SQUARE(childCount = 5, depth = 5);
+
+    override fun toString(): String = "$name(childCount=$childCount, depth=$depth)"
+  }
+
+  private val context = Unconfined
+
+  @Param
+  @JvmField var treeShape = TreeShape.SQUARE
+
+  private lateinit var workflow: FractalWorkflow
+  private lateinit var node: WorkflowNode<Props, Unit, *, Unit>
+
+  @Setup open fun setUp() {
+    println("Tree shape: $treeShape")
+    workflow = FractalWorkflow(treeShape.childCount, treeShape.depth)
+    node = workflow.createNode()
+  }
+
+  /**
+   * Always renders leaves. After the first render pass, the tree structure never changes, so this
+   * measures no-op render passes.
+   */
+  @Benchmark open fun renderPassAlwaysLeaves() {
+    node.render(workflow, RENDER_LEAVES)
+  }
+
+  /**
+   * Alternates between rendering leaves and not rendering leaves. The tree structure is always
+   * changing, as well as the props all the way down the tree, so this prevents props caching
+   * and measures setup/teardown cost for [WorkflowNode]s.
+   */
+  @Benchmark open fun renderPassAlternateLeaves() {
+    node.render(workflow, RENDER_LEAVES)
+    node.render(workflow, DO_NOT_RENDER_LEAVES)
+  }
+
+  /**
+   * Alternates between having every leaf run a worker, and having no leaves run workers.
+   * Measures setup/teardown cost for [Worker]s.
+   */
+  @Benchmark open fun renderPassAlternateWorkers() {
+    node.render(workflow, RENDER_LEAVES)
+    node.render(workflow, RUN_WORKERS)
+  }
+
+  private fun FractalWorkflow.createNode() = WorkflowNode(
+      id = this.id(),
+      workflow = this,
+      initialProps = RENDER_LEAVES,
+      snapshot = null,
+      baseContext = context
+  )
+}
+
+private class FractalWorkflow(
+  private val childCount: Int,
+  private val depth: Int
+) : StatefulWorkflow<Props, Unit, Nothing, Unit>() {
```

Comment:
Couldn't this be `StatelessWorkflow`?

Reply:
Yes, but I wanted to avoid the extra allocation.

Reply:
Will document.